### PR TITLE
feat: basic auth API

### PR DESCRIPTION
**Pull request by zhaohuabing**
Related to: https://github.com/envoyproxy/gateway/issues/1660

This PR adds API for HTTP Basic Auth:

```yaml
- apiVersion: gateway.envoyproxy.io/v1alpha1
  kind: SecurityPolicy
  metadata:
    name: basic-auth
    namespace: default
  spec:
    basicAuth:
      users:
        name: "htpasswd-secret"
    targetRef:
      group: gateway.networking.k8s.io
      kind: HTTPRoute
      name: httproute-1
```

### CODE REVIEW

File: api/v1alpha1/basic_auth_types.go [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+type BasicAuth struct {
+	// Username-hashed password pairs used to verify user credentials in the
+	// "Authorization" header.
+	//
+	// The value needs to be the htpasswd format, for example: "user1:{SHA}hashed_user1_password".
+	// Right now, ony SHA hash algorithm is supported.
+	// Reference to https://httpd.apache.org/docs/2.4/programs/htpasswd.html
+	//
+	// +kubebuilder:validation:MinItems=1
+	Users []string `json:"users"`
```
Comment (zhaohuabing) [inline api/v1alpha1/basic_auth_types.go:18]: We could store the user list in a k8s secret, but I am inclined to just use a plain string array for easier UI. The passwords are hashed, so it should be ok to put them in the SecurityPolicy resources.

Comment (arkodg) [inline api/v1alpha1/basic_auth_types.go:18]: must be a secret
we can either support multiple basic auth secrets https://kubernetes.io/docs/concepts/configuration/secret/#basic-authentication-secret which might be too much to type out and manage or support a generic secret with `data` containing the key `htpasswd` with its contents in it

Comment (zhaohuabing) [reply api/v1alpha1/basic_auth_types.go:18]: I vote to put `htpaswd` file in the `data` of a generic secret,  like the nginx k8s controller does: https://kubernetes.github.io/ingress-nginx/examples/auth/basic/

Comment (arkodg) [reply api/v1alpha1/basic_auth_types.go:18]: yeah sgtm, reg key naming, should it be `.htpasswd` to reflect the name of the file ?

Comment (zhaohuabing) [reply api/v1alpha1/basic_auth_types.go]: That's a good point.

File: api/v1alpha1/basic_auth_types.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import gwapiv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+// BasicAuth defines the configuration for 	the HTTP Basic Authentication.
+type BasicAuth struct {
+	// The Kubernetes secret which contains the username-password pairs used to
```
Comment (arkodg) [inline api/v1alpha1/basic_auth_types.go:12]:
```suggestion
	// The Kubernetes secret which contains the username-password pairs in htpasswd format, used to
```

**Comment by zhaohuabing** [comment]
@saltbo

File: api/v1alpha1/basic_auth_types.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+	// htpasswd format, used to verify user credentials in the "Authorization"
+	// header.
+	//
+	// This is an Opaque secret. The username-password pairs should be stored in
+	// the key ".htpasswd". As the key name indicates, the value needs to be the
+	// htpasswd format, for example: "user1:{SHA}hashed_user1_password".
+	// Right now, only SHA hash algorithm is supported.
+	// Reference to https://httpd.apache.org/docs/2.4/programs/htpasswd.html
+	// for more details.
+	Users gwapiv1b1.SecretObjectReference `json:"users"`
```
Comment (arkodg) [inline api/v1alpha1/basic_auth_types.go:22]: prefer if this field is called `Htpasswd`

**Comment by arkodg** [comment]
ptal @envoyproxy/gateway-maintainers

File: api/v1alpha1/basic_auth_types.go (hunk 3 above)
Comment (zhaohuabing) [reply api/v1alpha1/basic_auth_types.go:22]: I prefer `Users` because it's a user list, htpasswd is just the file format.

**Review by arkodg — APPROVED** [review]
LGTM thanks !

File: api/v1alpha1/basic_auth_types.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+import gwapiv1b1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+// BasicAuth defines the configuration for 	the HTTP Basic Authentication.
+type BasicAuth struct {
+	// The Kubernetes secret which contains the username-password pairs in
+	// htpasswd format, used to verify user credentials in the "Authorization"
+	// header.
+	//
+	// This is an Opaque secret. The username-password pairs should be stored in
```
Comment (arkodg) [inline api/v1alpha1/basic_auth_types.go:16]: maybe also add a line mentioning that secret must live in the same namespace as the policy

**Review by arkodg — APPROVED** [review]

**Review by zirain — APPROVED** [review]

**Pull request merged**